### ISSUE TITLE
TTS & STT Options ? 

### ISSUE THREAD

**Issue by m7mdcc** [felipem1210/freetalkbot#27]
Hi!

I’m really excited about this project! I have a similar one that uses JavaScript.

I would like to include an option for Speech-to-Text (STT). I’ve found that the Facebook model provides better results compared to Whisper-related models. You can check it out here: [Seamless M4T Large](https://huggingface.co/facebook/seamless-m4t-large).

For Text-to-Speech (TTS), I recommend using MsEdge. Here’s the link: [MsEdge TTS](https://github.com/Migushthe2nd/MsEdgeTTS).

What do you think about using Ollama with Flowise instead of Rasa AI?

**Comment by felipem1210** [comment]
Hi!, thanks for your interest

The SeamlessM4T Large model sounds very good. I saw the the v2 of this model makes both STT and TTS, and I find appropriate to add it as an alternative for using it in the VoIP channel. 
Now, to implement it, I see necessary to create an HTTP wrapper to access the model in python. Then we can dockerize it and use it in the docker-compose here. That is what faster-whisper-server is (for whisper). I would prefer not do it, as it would be a different repo, so if you have seen another project doing it please share it.

**Comment by m7mdcc** [comment]
I have experimented with both SeamlessM4T v1 and v2 using Python and WebSocket. Below is a sample of the code I've used. However, I'm not very familiar with real-time transcription and would like to receive results from the server as quickly as possible.



```
from fastapi import FastAPI, WebSocket, WebSocketDisconnect
from transformers import pipeline
import soundfile as sf
import torch
import torchaudio
import nltk
import traceback
nltk.download('punkt')

app = FastAPI()


device = "cuda:0"
saved_language = "arb"

# Initialize the speech recognition pipeline
asr_pipe = pipeline(
    "automatic-speech-recognition",
    model="seamless",
    device=device,
  #  chunk_length_s=30,
  #  batch_size=16,
    torch_dtype=torch.float16
)

# Initialize the translation pipeline
translation_pipe = pipeline(
    "translation",
    model="seamless",
    device=device,
    torch_dtype=torch.float16
)

@app.websocket("/ws/transcribe")
async def websocket_transcribe(websocket: WebSocket):
    await websocket.accept()
    try:
        while True:
            # Receive audio data chunk from WebSocket
            audio_chunk = await websocket.receive_bytes()

            # Save the chunk to a temporary file
            with open("temp_audio_chunk.wav", "wb") as f:
                f.write(audio_chunk)
            
            # Read and process the audio chunk
            audio, rate = sf.read("temp_audio_chunk.wav", dtype='float32')
            
            # Convert stereo to mono if necessary
            if len(audio.shape) > 1 and audio.shape[1] > 1:
                audio = torch.mean(torch.tensor(audio), dim=1).numpy()

            # Resample if necessary
            if rate != 16000:
                audio = torchaudio.functional.resample(torch.tensor(audio), rate, 16000).numpy()

            # Process with the ASR pipeline
            result = asr_pipe(audio, generate_kwargs={"tgt_lang": saved_language})

            # Send the transcription back to the client
            await websocket.send_text(result['text'])
    except WebSocketDisconnect:
        print("Client disconnected")
    except Exception as e:
        traceback.print_exc()
        await websocket.send_text(f"Error: {str(e)}")
```

this code working as expected but i would like to see result as it come from server

**Comment by felipem1210** [comment]
Basically what I do is to pass the complete data of the person voice once it finishes speaking. It is []byte type. 
https://github.com/felipem1210/freetalkbot/blob/main/packages/audiosocket/main.go#L134-L153
I don't think that can be catalogued as "real time streaming"  as I am waiting until the person finish talking to pass it completely to the websocket server. That was the way I found to do when using fast whisper server. 

Checking your code, I am seeing that you are generating a wav file from the audio chunk, but if I am not mistaken the idea is that the audio chunk be only a portion of what person talk so we can be generating the transcription by chunks and receiving the results by chunks for later concatenate them all and that way finish with the complete string?

**Comment by m7mdcc** [comment]
Yes, that's the current logic of the code. I need to update it to transcribe in real-time, meaning as soon as the person finishes talking, the previous segment is already transcribed. This way, there won't be any delay.

something like this  : https://github.com/Recap-Minutes-of-Meetings/whisper-transcribe-server/blob/main/src/server.py

**Comment by felipem1210** [comment]
Do you want to host your code in a repository and to create a docker image to it? so I can use add it as an alternative for STT. I would like also to be able to use the TTS capability of the model, so if you could generate another endpoint for TTS would be desirable!!

**Comment by felipem1210** [comment]
I think you can test it with this command:

`ffmpeg -loglevel quiet -f alsa -i default -ac 1 -ar 16000 -f s16le - | python yourcode.py`

With that you are capturing the data from your microphone and passing it as input for your server

**Comment by felipem1210** [comment]
@m7mdcc did you happen to test/publish your server?

**Comment by m7mdcc** [comment]
@felipem1210 Sorry not yet, but it's on progress . do you have time to arrange virtual meeting to discuss this ?

**Comment by felipem1210** [comment]
@m7mdcc you can add me in discord, my nickname is the same github username

**Comment by m7mdcc** [comment]
> @m7mdcc you can add me in discord, my nickname is the same github username

i can't find yours on discord , add me there : chunkyme
